File: cli-go/internal/domain/image_registry_test.go
package domain_test

import (
	"os"
	"testing"

	"github.com/adrg/xdg"
	"github.com/joacohbc/my-devcontainer-installer/cli-go/internal/domain"
)

func withTempRegistryDir(t *testing.T, fn func()) {
	t.Helper()
	tmp, err := os.MkdirTemp("", "dc-cli-registry-test-")
	if err != nil {
		t.Fatalf("failed to create temp dir: %v", err)
	}
	t.Cleanup(func() { os.RemoveAll(tmp) })

	prev, hasPrev := os.LookupEnv("XDG_CONFIG_HOME")
	os.Setenv("XDG_CONFIG_HOME", tmp)
	xdg.Reload()
	t.Cleanup(func() {
		if hasPrev {
			os.Setenv("XDG_CONFIG_HOME", prev)
		} else {
			os.Unsetenv("XDG_CONFIG_HOME")
		}
		xdg.Reload()
	})

	fn()
}

func TestComputeFingerprint_sameContentProducesSameResult(t *testing.T) {
	fp1 := domain.ComputeFingerprint("FROM ubuntu:24.04\nRUN apt-get update", nil, []string{"base"})
	fp2 := domain.ComputeFingerprint("FROM ubuntu:24.04\nRUN apt-get update", nil, []string{"base"})
	if fp1 != fp2 {
		t.Errorf("expected identical fingerprints, got %q and %q", fp1, fp2)
	}
}

func TestComputeFingerprint_labelChangeDoesNotAffectResult(t *testing.T) {
	withoutLabel := domain.ComputeFingerprint("FROM ubuntu:24.04\nRUN apt-get update", nil, nil)
	withLabel := domain.ComputeFingerprint("FROM ubuntu:24.04\nRUN apt-get update\nLABEL foo=\"bar\"", nil, nil)
	if withoutLabel != withLabel {
		t.Errorf("LABEL change should not affect fingerprint: %q vs %q", withoutLabel, withLabel)
	}
}

func TestComputeFingerprint_multilineLabelStripped(t *testing.T) {
	withoutLabel := domain.ComputeFingerprint("FROM ubuntu:24.04", nil, nil)
	withMultilineLabel := domain.ComputeFingerprint("FROM ubuntu:24.04\nLABEL foo=\"bar\" \\\n      baz=\"qux\"", nil, nil)
	if withoutLabel != withMultilineLabel {
		t.Errorf("multiline LABEL should be stripped: %q vs %q", withoutLabel, withMultilineLabel)
	}
}

func TestComputeFingerprint_differentDockerfileProducesDifferentResult(t *testing.T) {
	fp1 := domain.ComputeFingerprint("FROM ubuntu:24.04", nil, nil)
	fp2 := domain.ComputeFingerprint("FROM ubuntu:22.04", nil, nil)
	if fp1 == fp2 {
		t.Error("different Dockerfiles should produce different fingerprints")
	}
}

func TestComputeFingerprint_differentCopyFilesProducesDifferentResult(t *testing.T) {
	files1 := map[string]string{"script.sh": "echo hello"}
	files2 := map[string]string{"script.sh": "echo world"}
	fp1 := domain.ComputeFingerprint("FROM ubuntu:24.04", files1, nil)
	fp2 := domain.ComputeFingerprint("FROM ubuntu:24.04", files2, nil)
	if fp1 == fp2 {
		t.Error("different copy file contents should produce different fingerprints")
	}
}

func TestComputeFingerprint_differentModuleIDsProducesDifferentResult(t *testing.T) {
	fp1 := domain.ComputeFingerprint("FROM ubuntu:24.04", nil, []string{"base", "nodejs"})
	fp2 := domain.ComputeFingerprint("FROM ubuntu:24.04", nil, []string{"base", "python"})
	if fp1 == fp2 {
		t.Error("different module IDs should produce different fingerprints")
	}
}

func TestComputeFingerprint_moduleIDOrderDoesNotMatter(t *testing.T) {
	fp1 := domain.ComputeFingerprint("FROM ubuntu:24.04", nil, []string{"nodejs", "base"})
	fp2 := domain.ComputeFingerprint("FROM ubuntu:24.04", nil, []string{"base", "nodejs"})
	if fp1 != fp2 {
		t.Errorf("module ID order should not affect fingerprint: %q vs %q", fp1, fp2)
	}
}

func TestFingerprintTag_usesFirst12Chars(t *testing.T) {
	fp := "abcdef123456789012"
	tag := domain.FingerprintTag(fp)
	expected := "devcontainer-cli/abcdef123456:latest"
	if tag != expected {
		t.Errorf("expected %q, got %q", expected, tag)
	}
}

func TestFingerprintTag_shortFingerprintUsedWhole(t *testing.T) {
	fp := "abc"
	tag := domain.FingerprintTag(fp)
	expected := "devcontainer-cli/abc:latest"
	if tag != expected {
		t.Errorf("expected %q, got %q", expected, tag)
	}
}

func TestRecordEntry_addsNewEntry(t *testing.T) {
	withTempRegistryDir(t, func() {
		entry := domain.ImageEntry{
			ProjectDir:  "/home/user/myproject",
			Workspace:   "myproject",
			Mode:        "local-cached",
			Image:       "devcontainer-cli/abc123:latest",
			Fingerprint: "abc123",
			CreatedAt:   "2024-01-01T00:00:00Z",
			LastUpdated: "2024-01-01T00:00:00Z",
		}
		if err := domain.RecordEntry(entry); err != nil {
			t.Fatalf("RecordEntry failed: %v", err)
		}

		entries := domain.ListEntries()
		if len(entries) != 1 {
			t.Fatalf("expected 1 entry, got %d", len(entries))
		}
		if entries[0].ProjectDir != entry.ProjectDir {
			t.Errorf("expected project dir %q, got %q", entry.ProjectDir, entries[0].ProjectDir)
		}
	})
}

func TestRecordEntry_updatesExistingEntryByProjectDir(t *testing.T) {
	withTempRegistryDir(t, func() {
		first := domain.ImageEntry{
			ProjectDir:  "/home/user/myproject",
			Workspace:   "myproject",
			Mode:        "local-cached",
			Image:       "devcontainer-cli/aaa:latest",
			Fingerprint: "aaa",
			CreatedAt:   "2024-01-01T00:00:00Z",
			LastUpdated: "2024-01-01T00:00:00Z",
		}
		if err := domain.RecordEntry(first); err != nil {
			t.Fatalf("first RecordEntry failed: %v", err)
		}

		updated := first
		updated.Image = "devcontainer-cli/bbb:latest"
		updated.Fingerprint = "bbb"
		updated.LastUpdated = "2024-02-01T00:00:00Z"
		if err := domain.RecordEntry(updated); err != nil {
			t.Fatalf("second RecordEntry failed: %v", err)
		}

		entries := domain.ListEntries()
		if len(entries) != 1 {
			t.Fatalf("expected 1 entry after update, got %d", len(entries))
		}
		if entries[0].Image != updated.Image {
			t.Errorf("expected updated image %q, got %q", updated.Image, entries[0].Image)
		}
	})
}

func TestFindByFingerprint_returnsMatchingEntry(t *testing.T) {
	withTempRegistryDir(t, func() {
		entry := domain.ImageEntry{
			ProjectDir:  "/home/user/proj",
			Workspace:   "proj",
			Mode:        "local-cached",
			Image:       "devcontainer-cli/fp12345:latest",
			Fingerprint: "fp12345",
			CreatedAt:   "2024-01-01T00:00:00Z",
			LastUpdated: "2024-01-01T00:00:00Z",
		}
		_ = domain.RecordEntry(entry)

		found := domain.FindByFingerprint("fp12345")
		if found == nil {
			t.Fatal("expected to find entry by fingerprint, got nil")
		}
		if found.ProjectDir != entry.ProjectDir {
			t.Errorf("expected project dir %q, got %q", entry.ProjectDir, found.ProjectDir)
		}
	})
}

func TestFindByFingerprint_returnsNilWhenMissing(t *testing.T) {
	withTempRegistryDir(t, func() {
		found := domain.FindByFingerprint("doesnotexist")
		if found != nil {
			t.Errorf("expected nil, got entry %+v", found)
		}
	})
}

func TestRemoveEntry_removesExistingEntry(t *testing.T) {
	withTempRegistryDir(t, func() {
		entry := domain.ImageEntry{
			ProjectDir:  "/home/user/proj",
			Workspace:   "proj",
			Mode:        "local-cached",
			Image:       "img:latest",
			CreatedAt:   "2024-01-01T00:00:00Z",
			LastUpdated: "2024-01-01T00:00:00Z",
		}
		_ = domain.RecordEntry(entry)

		removed := domain.RemoveEntry("/home/user/proj")
		if !removed {
			t.Error("expected RemoveEntry to return true")
		}

		entries := domain.ListEntries()
		if len(entries) != 0 {
			t.Errorf("expected 0 entries after removal, got %d", len(entries))
		}
	})
}

func TestRemoveEntry_returnsFalseWhenNotFound(t *testing.T) {
	withTempRegistryDir(t, func() {
		removed := domain.RemoveEntry("/nonexistent/path")
		if removed {
			t.Error("expected RemoveEntry to return false for nonexistent entry")
		}
	})
}

File: cli/src/__tests__/image-registry.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import {
  computeFingerprint,
  fingerprintTag,
  loadRegistry,
  recordEntry,
  removeEntry,
  findByFingerprint,
  saveRegistry,
  imageRegistryPath,
} from '@/domain/image-registry.js';

function withTempHome<T>(fn: () => T): T {
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'dc-cli-imgreg-'));
  const origXdg = process.env.XDG_CONFIG_HOME;
  const origAppData = process.env.APPDATA;
  process.env.XDG_CONFIG_HOME = tmp;
  process.env.APPDATA = tmp;
  try {
    return fn();
  } finally {
    process.env.XDG_CONFIG_HOME = origXdg;
    process.env.APPDATA = origAppData;
    fs.rmSync(tmp, { recursive: true, force: true });
  }
}

const DF = `FROM ubuntu:24.04
RUN apt-get update
LABEL foo=bar
LABEL multi="line" \\
  more=stuff
RUN echo hello
`;

test('computeFingerprint stable for identical inputs', () => {
  const a = computeFingerprint(DF, { 'a.sh': 'echo a' }, ['nodejs', 'go']);
  const b = computeFingerprint(DF, { 'a.sh': 'echo a' }, ['go', 'nodejs']);
  assert.equal(a, b);
});

test('computeFingerprint ignores LABEL-only diffs', () => {
  const a = computeFingerprint(DF, {}, []);
  const stripped = DF.replace('LABEL foo=bar', 'LABEL foo=baz');
  const b = computeFingerprint(stripped, {}, []);
  assert.equal(a, b);
});

test('computeFingerprint changes when copyFile content changes', () => {
  const a = computeFingerprint(DF, { 'a.sh': 'echo a' }, []);
  const b = computeFingerprint(DF, { 'a.sh': 'echo DIFFERENT' }, []);
  assert.notEqual(a, b);
});

test('computeFingerprint changes when Dockerfile body changes', () => {
  const a = computeFingerprint(DF, {}, []);
  const b = computeFingerprint(DF + '\nRUN echo extra', {}, []);
  assert.notEqual(a, b);
});

test('fingerprintTag uses first 12 hex chars', () => {
  const t = fingerprintTag('0123456789abcdef0123456789abcdef');
  assert.equal(t, 'devcontainer-cli/0123456789ab:latest');
});

test('recordEntry upserts by projectDir', () => {
  withTempHome(() => {
    const e = {
      projectDir: '/tmp/proj-1',
      workspace: 'proj1',
      mode: 'remote' as const,
      image: 'ghcr.io/x/y:latest',
      createdAt: '2026-05-22T00:00:00Z',
      lastUpdated: '2026-05-22T00:00:00Z',
    };
    recordEntry(e);
    recordEntry({ ...e, lastUpdated: '2026-05-22T01:00:00Z' });
    const entries = loadRegistry();
    assert.equal(entries.length, 1);
    assert.equal(entries[0].lastUpdated, '2026-05-22T01:00:00Z');
  });
});

test('removeEntry no-ops on missing', () => {
  withTempHome(() => {
    assert.equal(removeEntry('/nonexistent'), false);
    recordEntry({
      projectDir: '/tmp/p',
      workspace: 'p',
      mode: 'local-cached',
      image: 'p:local',
      createdAt: 'now',
      lastUpdated: 'now',
    });
    assert.equal(removeEntry('/tmp/p'), true);
    assert.equal(loadRegistry().length, 0);
  });
});

test('findByFingerprint returns matching entry', () => {
  withTempHome(() => {
    recordEntry({
      projectDir: '/tmp/a',
      workspace: 'a',
      mode: 'local-cached',
      image: 'devcontainer-cli/abc123:latest',
      fingerprint: 'abc123def456',
      createdAt: 'now',
      lastUpdated: 'now',
    });
    const hit = findByFingerprint('abc123def456');
    assert.ok(hit);
    assert.equal(hit.workspace, 'a');
    assert.equal(findByFingerprint('nope'), undefined);
  });
});

test('loadRegistry returns [] when file missing', () => {
  withTempHome(() => {
    assert.deepEqual(loadRegistry(), []);
  });
});

test('saveRegistry creates parent directory', () => {
  withTempHome(() => {
    saveRegistry([
      {
        projectDir: '/x',
        workspace: 'x',
        mode: 'remote',
        image: 'i',
        createdAt: 'now',
        lastUpdated: 'now',
      },
    ]);
    assert.ok(fs.existsSync(imageRegistryPath()));
  });
});
